File: js/12015.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const N = +inputs[0];
const arr = inputs[1].split(' ').map(Number);

const lis = [arr[0]];

for (let i = 1; i < N; i++) {
  if (lis[lis.length - 1] < arr[i]) {
    lis.push(arr[i]);
  } else {
    const idx = binarySearch(0, lis.length - 1, arr[i]);
    lis[idx] = arr[i];
  }
}

console.log(lis.length);

function binarySearch(start, end, target) {
  while (start < end) {
    const mid = Math.floor((start + end) / 2);

    if (lis[mid] < target) start = mid + 1;
    else end = mid;
  }

  return end;
}

File: c++/12015.cpp
#include <iostream>
#include <vector>

using namespace std;

int arr[1000000];
vector<int> lis;

int binarySearch(int start, int end, int target) {
  while (start < end) {
    int mid = (start + end) / 2;

    if (lis[mid] < target) {
      start = mid + 1;
    } else {
      end = mid;
    }
  }
  return end;
}

int main() {
  freopen("example.txt", "r", stdin);
  int N;
  cin >> N;

  for (int i = 0; i < N; i++) {
    cin >> arr[i];
  }

  lis.push_back(arr[0]);

  for (int i = 1; i < N; i++) {
    if (lis.back() < arr[i]) {
      lis.push_back(arr[i]);
    } else {
      int idx = binarySearch(0, lis.size() - 1, arr[i]);
      lis[idx] = arr[i];
    }
  }

  cout << lis.size();
  return 0;
}